src/Identity/Core/src/IdentityApiEndpointRouteBuilderExtensions.cs

```diff
@@ -72,6 +75,27 @@ public static IEndpointConventionBuilder MapIdentityApi<TUser>(this IEndpointRou
             return TypedResults.SignIn(claimsPrincipal, authenticationScheme: scheme);
         });
 
+        routeGroup.MapPost("/refresh", async Task<Results<UnauthorizedHttpResult, Ok<AccessTokenResponse>, SignInHttpResult, ChallengeHttpResult>>
+            ([FromBody] RefreshRequest refreshRequest, [FromServices] IOptionsMonitor<BearerTokenOptions> optionsMonitor, [FromServices] TimeProvider timeProvider, [FromServices] IServiceProvider sp) =>
+        {
+            var signInManager = sp.GetRequiredService<SignInManager<TUser>>();
+            var identityBearerOptions = optionsMonitor.Get(IdentityConstants.BearerScheme);
+            var refreshTokenProtector = identityBearerOptions.RefreshTokenProtector ?? throw new ArgumentException($"{nameof(identityBearerOptions.RefreshTokenProtector)} is null", nameof(optionsMonitor));
+            var refreshTicket = refreshTokenProtector.Unprotect(refreshRequest.RefreshToken);
+
+            // Reject the /refresh attempt with a 401 if the token expired or the security stamp validation fails
+            if (refreshTicket?.Properties?.ExpiresUtc is not { } expiresUtc ||
+                timeProvider.GetUtcNow() >= expiresUtc ||
+                await signInManager.ValidateSecurityStampAsync(refreshTicket.Principal) is not TUser user)
```

Comment:
I think this is a good idea, but I'm going to leave it as a follow up. #48947

One problem with this is that we would have to regenerate the access token anyway since it's not sent as part of this request. It wouldn't be as much of an issue to recreate if we had the IssuedUtc and not just the expiration. And even if we did have this, you could just keep using an old refresh token to force a db hit, so it'd only defend against buggy clients not malicious ones.

Reply:
There are a lot more buggy clients than malicious ones :->